### ISSUE TITLE
What is the best way to handle configuration migration?

### ISSUE THREAD

**Issue by philpax** [goatcorp/faq#60]
Sometimes, you just gotta change how your config's being stored / what's being stored, but you want to make sure that you aren't going to lose user data while doing so. What's the cleanest way of doing this, preferably while remaining a nice and clean separation between versions?

This seems like it might be a bit complicated because the `Configuration` is automatically de/serialised - how do you gracefully choose to deserialise with a ConfigurationV1, convert to a ConfigurationV2, and serialise with the ConfigurationV2?

**Comment by karashiiro** [comment]
See [here](https://github.com/karashiiro/PingPlugin/blob/main/PingPlugin/PingConfiguration.cs).

**Comment by karashiiro** [comment]
Also note all of the obsolete fields required for proper loading.

**Comment by philpax** [comment]
Yeah, that's all I can think of as well with the current system (outside of like... having explicit `v1: ConfigurationV1` fields). Bit of a shame.
